overlord/snapstate/handlers.go

```diff
@@ -1147,9 +1148,43 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		}
 		return copyDataErr
 	}
+
+	// the migration hasn't been done - do it now
+	if opts.UseHidden && !opts.MigratedToHidden {
+		if err := m.migrateToHiddenDir(t); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+func (m *SnapManager) migrateToHiddenDir(t *state.Task) error {
+	t.State().Lock()
+	snapsup, err := TaskSnapSetup(t)
+	t.State().Unlock()
+	if err != nil {
+		return err
+	}
+
+	// set the migrated status even on error, so we undo this later
+	err = m.backend.HideSnapData(snapsup.InstanceName())
+	snapsup.MigratedHidden = true
```

Comment:
is there a reason to keep this of snapsup as well instead of changing SnapState already here?

Reply:
It allows us to distinguish between a migration done in this run and in a previous run, which is useful when undoing. I saw the same pattern applied to other things as well, doLinkSnap copies a lot of fields from SnapSetup to SnapState and then sets it a the end when it can't fail anymore